neo/IO/Caching/DataCache.cs

```diff
@@ -78,137 +84,159 @@ public void Commit()
 
         public void Delete(TKey key)
         {
-            if (dictionary.TryGetValue(key, out Trackable trackable))
+            lock (dictionary)
             {
-                if (trackable.State == TrackState.Added)
-                    dictionary.Remove(key);
+                if (dictionary.TryGetValue(key, out Trackable trackable))
+                {
+                    if (trackable.State == TrackState.Added)
+                        dictionary.Remove(key);
+                    else
+                        trackable.State = TrackState.Deleted;
+                }
                 else
-                    trackable.State = TrackState.Deleted;
-            }
-            else
-            {
-                TValue item = TryGetInternal(key);
-                if (item == null) return;
-                dictionary.Add(key, new Trackable
                 {
-                    Key = key,
-                    Item = item,
-                    State = TrackState.Deleted
-                });
+                    TValue item = TryGetInternal(key);
+                    if (item == null) return;
+                    dictionary.Add(key, new Trackable
+                    {
+                        Key = key,
+                        Item = item,
+                        State = TrackState.Deleted
+                    });
+                }
             }
         }
 
         public abstract void DeleteInternal(TKey key);
 
         public void DeleteWhere(Func<TKey, TValue, bool> predicate)
         {
-            foreach (Trackable trackable in dictionary.Where(p => p.Value.State != TrackState.Deleted && predicate(p.Key, p.Value.Item)).Select(p => p.Value))
-                trackable.State = TrackState.Deleted;
+            lock (dictionary)
+            {
+                foreach (Trackable trackable in dictionary.Where(p => p.Value.State != TrackState.Deleted && predicate(p.Key, p.Value.Item)).Select(p => p.Value))
+                    trackable.State = TrackState.Deleted;
+            }
         }
 
         public IEnumerable<KeyValuePair<TKey, TValue>> Find(byte[] key_prefix = null)
         {
-            foreach (var pair in FindInternal(key_prefix ?? new byte[0]))
-                if (!dictionary.ContainsKey(pair.Key))
-                    yield return pair;
-            foreach (var pair in dictionary)
-                if (pair.Value.State != TrackState.Deleted && (key_prefix == null || pair.Key.ToArray().Take(key_prefix.Length).SequenceEqual(key_prefix)))
-                    yield return new KeyValuePair<TKey, TValue>(pair.Key, pair.Value.Item);
+            lock (dictionary)
```

Comment:
What about this particular code path? This does not write to the dictionary. Would it be advantageous to allow multiple readers simultaneously through this code path?

Reply:
This `Find` method is rarely used.